omnibus/files/private-chef-cookbooks/private-chef/libraries/ec_postgres.rb

```diff
@@ -1,14 +1,29 @@
 class EcPostgres
+  # Provides a superuser connection to the specified database
   def self.with_connection(node, database = 'template1')
-    require 'pg'
-    as_user(node['private_chef']['postgresql']['username']) do
-      connection = ::PGconn.open('dbname' => database)
-      begin
-        yield connection
-      ensure
-        connection.close
-      end
-    end
+     require 'pg'
+     postgres = node['private_chef']['postgresql']
+     if postgres['external']
+         connection = ::PGconn.open('user' => postgres['db_superuser'],
+                                    'host' => postgres['vip'], 'password' => postgres['db_superuser_password'],
+                                    'port' => postgres['port'], 'dbname' => database)
+         begin
+           yield connection
+         ensure
+           connection.close
+         end
+     else
+       # Local administrative connections are still done using the
+       # dedicated postgresql user over pipe.
+       as_user(postgres['username']) do
+         connection = ::PGconn.open('dbname' => database)
+         begin
+           yield connection
+         ensure
+           connection.close
+         end
+       end
+     end
```

Comment:
Will something like this work to reduce the duplication here? Warn: Untested code written in comment box:

```ruby
def self.with_connection(node, database = 'template1')
  require 'pg'
  postgres = node['private_chef']['postgresql']
  connection = if postgres['external']
                 ::PGconn.open('user' => postgres['db_superuser'],
                               'password' => postgres['db_superuser_password'],
                               'host' => postgres['vip'], 
                               'port' => postgres['port'],
                               'dbname' => database)
               else
                 as_user(postgres['username']) do
                   ::PGconn.open('dbname' => database)
                 end
               end
  begin
    yield connection
  ensure
    connection.close
  end
end
```